DOCS-9684-X12-template (#1179)
mulesoft/docs-connectors: x12-edi/2.5/modules/ROOT/pages/index.adoc

= X12 EDI Connector 2.5.0 - Mule 4
:page-aliases: connectors::x12-edi/x12-edi-connector.adoc

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Premium] +

> Anypoint Connector for X12 EDI lets you convert X12 messages to and from DataWeave-compatible representations using lists and maps.

Reviewer: ```suggestion
Anypoint Connector for X12 EDI (X12 Connector) enables you to convert X12 messages to and from DataWeave-compatible representations using lists and maps.
```

> The schema definitions for all https://docs.mulesoft.com/connectors/x12-edi/x12-edi-versions-x12[supported X12 standard transaction sets] are provided.

Reviewer: This is the wrong way to cross reference. It should be the standard xref format

Many HIPAA document versions are also supported, as listed separately in the https://docs.mulesoft.com/connectors/x12-edi/x12-edi-versions-hipaa[Connector HIPAA reference page].

== Prerequisites

To use this connector, you must be familiar with:

* Anypoint Connectors
* Mule runtime engine (Mule)
* Elements and global elements in a Mule flow
* Creating a Mule app using Anypoint Studio (Studio)

> * To use the X12 EDI connector in a production environment, you must

Reviewer: ```suggestion
To use the X12 EDI connector in a production environment, you must have:
```

> have purchased a MuleSoft license for Anypoint B2B.

Reviewer: A MuleSoft license for Anypoint B2B  I don't think this is correct, is it?  Can you verify that there is such a thing.

> * Anypoint Studio version <7.0> or later

Reviewer: ```suggestion
* Anypoint Studio 7.0 or later
```

> NOTE: X12 EDI Connector supports files of up to 15 MB, and the memory requirement is roughly 40 to 1.

Reviewer: ```suggestion
[NOTE]
X12 Connector supports files of up to 15 MB in size, and the memory requirement is approximately 40 to 1.
```

Reviewer: That memory requirement doesn't make sense. 40 to 1?  Is it 1 - 40 MB?

> For example, a 1-MB file requires up to 40 MB of memory to process, so it's important to consider this memory requirement in conjunction with your TPS needs for large X12 files.

Reviewer: TPS needs? We probably should spell out first use.

> This is not an exact figure; the value might vary according to the complexity of the mapping instructions.

Reviewer: ```suggestion
This is not an exact figure; the value varies according to the complexity of the mapping instructions.
```

> // == Common Use Cases for the Connector - there are currently no use cases for this connector.

Reviewer: You can remove anything that's commented out and that you aren't going to use.

** xref:<connector-name>-connector-config-topics.adoc[Additional Configuration Information]
to help extend your understanding of the <connector-name> connector
** xref:<connector-name>-connector-examples.adoc[Examples], which provide one or more use cases for the connector
* Advanced users, read:
** xref:<connector-name>-connector-xml-maven.adoc[XML and Maven Support]
** xref:<connector-name>-connector-config-topics.adoc[Additional Configuration Information] to help extend your understanding of the <connector-name> connector
** xref:<connector-name>-connector-examples.adoc[Examples], which provides one or more use cases for the connector

// The following line is only needed for connectors with upgrade information
// For a major or minor release, see
// xref:<connector-name>-connector-upgrade-migrate.adoc[Upgrade and Migrate].

== About the EDI Schema Language

The X12 EDI Connector uses a YAML format called ESL (for EDI Schema Language) to represent EDI schemas.
Basic ESLs define the structure of EDI messages in terms of structures (known as transaction sets in X12 terminology), groups (known as loops, in X12), segments, composites, and elements.

> ESLs for the all https://docs.mulesoft.com/connectors/x12-edi/x12-edi-versions-x12[supported X12 versions] are included.

Reviewer: Need to properly xref

If you are using HIPAA, then a set of https://docs.mulesoft.com/connectors/x12-edi/x12-edi-versions-hipaa[HIPAA variations] of base X12 transaction sets are included.

You can customize the X12 schema definitions to suit your data by defining an overlay schema. An overlay schema is a special form of ESL that allows you to modify a base schema, such as an X12 4010 850 schema, with your specific conventions.
You don't need an overlay schema if you're using the structure defined by the standard, but most EDI exchanges modify the base definitions in an implementation convention.

You can also define your own schemas from scratch, or copy and edit a base X12 schema definition to suit your data as seen in the EDI Schema Language Reference.

=== Readability
YAML uses a combination of lists and sets of key-value pairs. The order of
values is not important, as long as the required items are present. Quotes

> (such as single or double) are used around values that can consist

Reviewer: ```suggestion
(such as single or double) are used around values that consist
```

of digits but are meant to be interpreted as strings (otherwise the
YAML parser treats the values as numbers). Indentation is used to show the nesting of lists.

> The ESL structures shown here define all simple key-value pairs before any lists that are part of the same definition.

Reviewer: ```suggestion
The ESL structures shown here defines all simple key-value pairs before any lists that are part of the same definition.
```

== Set Your Schema Locations

To use the connector, you need to know the locations of the schemas
in your project. If you're using out-of-the-box X12 schemas and
not customizing anything, the schema location follows the
`/x12/{version}/{transaction set}.esl` convention.
For example, using the 005010 version and the 850 transaction set, has a schema location of `/x12/005010/850.esl`.

If you're using one or more custom schemas, locate these under
a directory in `src/main/app` and refer to the location relative to this directory.
For example, if an 850 schema (either overlay or full) is under `src/main/app/mypartner/850.esl,
the schema location is `/mypartner/850.esl`.

You can configure schema locations in the Anypoint Studio XML view. The schema files are loaded when the connector is initialized,
so runtime expressions such as flow variables are not supported.

> To configure schemas in Anypoint Studio:

Reviewer: There shouldn't be any tasks in the index file. This should be in a configuration topic.

> = X12 EDI Connector 2.5.0 - Mule 4

Reviewer: ```suggestion
= X12 EDI Connector 2.5 - Mule 4
```

Reviewer: Don't use x.x.x format for versions. We only update doc for minor and major releases, not patch releases.

:page-aliases: connectors::x12-edi/x12-edi-connector.adoc

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Premium] +

> Anypoint Connector for X12 EDI enables you to convert X12 messages to and from DataWeave-compatible representations using lists and maps.

Reviewer: ```suggestion
Anypoint Connector for X12 EDI (X12 Connector) enables you to convert X12 messages to and from DataWeave-compatible representations using lists and maps.
```

The schema definitions for all xref:x12-edi-versions-x12.adoc[supported X12 standard transaction sets] are provided.

Many HIPAA document versions are also supported, as listed separately in the xref:x12-edi-versions-hipaa.adoc[Connector HIPAA reference page].

== Prerequisites

To use this connector, you must be familiar with:

* Anypoint Connectors
* Mule runtime engine (Mule)
* Elements and global elements in a Mule flow
* Creating a Mule app using Anypoint Studio (Studio)

+ To use the X12 EDI connector in a production environment, you must have:

> * purchased a MuleSoft license for Anypoint B2B.

Reviewer: ```suggestion
* Purchased a MuleSoft license for Anypoint B2B
```

* Anypoint Studio 7.0 or later

[NOTE]

> X12 Connector supports files of up to 15 MB in size, and the memory requirement is approximately 40 to 1.

Reviewer: What does "40 to 1" mean?

For example, a 1-MB file requires up to 40 MB of memory to process, so it's important to consider this memory requirement in conjunction with your Transactions Per Second (TPS)  needs for large X12 files.
This is not an exact figure; the value varies according to the complexity of the mapping instructions.

// == Common Use Cases for the Connector - there are currently no use cases for this connector.

// Make the title singular if there is just one use case
// Add a lead in sentence and then list common use cases for the connector
// Make the title singular if there is just one case

// For examples, see xref:<connector-name>-connector-examples.adoc[Examples].

== About the EDI Schema Language

> The X12 EDI Connector uses a YAML format called ESL (for EDI Schema Language) to represent EDI schemas.

Reviewer: ```suggestion
X12 Connector uses a YAML format called ESL (for EDI Schema Language) to represent EDI schemas.
```

For example, if a message has the value `024020`, its DataWeave mapping
appears as `9620000` milliseconds.

The value `024020` is in the format "HHMMSS", so to convert it to milliseconds, use this formula:

----
(02*3600000) + (40*60000) + (20*1000) = 9620000
----

> == Using Exchange Templates and Examples

Reviewer: ```suggestion
== Exchange Templates and Examples
```

Reviewer: ^^That was a recent (like today) change to the template

https://www.mulesoft.com/exchange/[Anypoint Exchange] provides templates
that you can use as starting points for your apps and examples that illustrate a complete solution.

////
List and link to up to 10 Exchange templates and examples.
Use the Integration Pattern categories (broadcast, migration, bidirectional sync, aggregation).
////

== Next Step

> After you complete the prerequisites, you are ready to create your own app and configure the connector using xref:<connector-name>-connector-studio.adoc[Anypoint Studio].

Reviewer: ```suggestion
After you complete the prerequisites, you are ready to create your own app and configure the connector using xref:x12-edi-connector-studio.adoc[Anypoint Studio].
```

== See Also
////

> xref:connectors::introduction/introduction-to-anypoint-connectors.adoc[Introduction to Anypoint Connectors]

Reviewer: This xref shouldn't be commented out
